File: wpca.py
"""WPCA script in python."""

# Note: INPUT PARAMETERS
# _X_RAW_DB, _LO_LIMIT_1
# raw data in csv using db structure, i.e., a single row contains:
# code of representative, code of division, encoded vote (i.e. one of -1, 1, 0, NA)
# for example:
# “Joe Europe”,”Division-007”,”yes”

import pandas as pd
import numpy as np

localpath = "./"

# quick fix setting rotation
rotate = {
  'column': 'voter_id',
  'value': 6074,  # Petr Fiala
  'dims': [1, 1]
}

# quick settings for correct dates in charts
first_half_year = {"half": 2, "year": 2021}

Xsource = pd.read_csv(localpath + "data/votes.csv")

# recode options
conditions = [
  Xsource['option'].eq('yes'),
  Xsource['option'].isin(['no', 'abstain']),
  Xsource['option'].isin(['not voting', 'absent'])
]

nvalues = [1, -1, pd.NA]

Xsource['option_numeric'] = np.select(conditions, nvalues, default=pd.NA)

#Xrawdb = _X_RAW_DB
Xrawdb = Xsource

# lower limit to eliminate from calculations, e.g., .1; number
lo_limit = 0.1

Xraw = pd.pivot_table(Xrawdb, values='option_numeric', columns='voter_id', index='vote_event_id', aggfunc='first', fill_value=np.nan)

Xpeople = Xraw.columns
Xvote_events = Xraw.index

# Scale the data
Xstand = Xraw.sub(Xraw.mean(axis=1), axis=0).div(Xraw.std(axis=1, ddof=0), axis=0)

# WEIGHTS
# weights 1 for divisions, based on number of persons in division
w1 = (np.abs(Xraw) == 1).sum(axis=1, skipna=True) / (np.abs(Xraw) == 1).sum(axis=1, skipna=True).max()
w1[np.isnan(w1)] = 0

# weights 2 for divisions, "100:100" vs. "195:5"
w2 = 1 - np.abs((Xraw == 1).sum(axis=1, skipna=True) - (Xraw == -1).sum(axis=1, skipna=True)) / (~Xraw.isna()).sum(axis=1, skipna=True)
w2[np.isnan(w2)] = 0

# weighted scaled matrix; divisions x persons
X = Xstand.mul(w1, axis=0).mul(w2, axis=0)

# weighted scaled with NA substituted by 0; division x persons
X0 = X.fillna(0)

# filter people with too low attendance
w = (w1 * w2).sum()
pw = Xraw.notna().mul(w1, axis=0).mul(w2, axis=0).sum(axis=0) / w
selected_voters = pw[pw > lo_limit].index.tolist()
X0c = X0.loc[:, selected_voters]    # for direct use
X0c = X0    # for use with time intervals

# I matrix
I = X.notna().astype(int).loc[:, selected_voters]
Iw = I.fillna(0).mul(w1, axis=0).mul(w2, axis=0)

# “X’X” MATRIX
# weighted X’X matrix with missing values substituted and excluded persons; persons x persons
# C=t(X0c)%*%X0c * 1/(t(Iwc)%*%Iwc) * (sum(w1*w1*w2*w2))
C = (X0c.T.dot(X0c) * (1 / (Iw.T.dot(Iw))) * (w1 * w1 * w2 * w2).sum()).fillna(0)

# DECOMPOSITION
# eigendecomposition
eigvals, eigvecs = np.linalg.eig(C)
# sort the eigenvectors and eigenvalues
idx = eigvals.argsort()[::-1]
eigvals = eigvals[idx]
eigvecs = eigvecs[:, idx]

# projected divisions into dimensions
Xy = X0c.dot(eigvecs)

# lambda matrix
sigma = np.sqrt(eigvals)
sigma = np.nan_to_num(sigma, nan=0)
lmbda = np.diag(sigma)
# unit scaled lambda matrix
lambdau = np.sqrt(lmbda.dot(lmbda) / lmbda.dot(lmbda).sum())

# projection of persons into dimensions
Xproj = eigvecs.dot(lmbda)
# scaled projection of persons into dimensions
Xproju = eigvecs.dot(lambdau) * np.sqrt(len(eigvecs))
Xprojudf = pd.DataFrame(Xproju)
Xprojudf.index = selected_voters

# lambda^-1 matrix
lambda_1 = np.diag(np.sqrt(1 / eigvals))
lambda_1 = np.nan_to_num(lambda_1)

# Z (rotation values of divisions)
Z = X0c.dot(eigvecs).dot(lambda_1)

# second projection
Xproj2 = X0c.T.dot(Z)
# without missing values, they are equal:
# Xproj, Xproj2

# save Xproju
out = Xprojudf.iloc[:, range(0, 3)]
out.index = Xproj2.index
out.reset_index(inplace=True)
out.columns = ['voter_id', 'dim1', 'dim2', 'dim3']

# rotate
row = out.loc[out[rotate['column']] == rotate['value']]
# dims
for i in range(0, 2):
  if (rotate['dims'][i]) * row['dim' + str(i + 1)].values[0] < 0:
    out.loc[:, 'dim' + str(i + 1)] = out['dim' + str(i + 1)] * -1

out.to_csv(localpath + "data/wpca.csv", index=False)



# PROJECTIONS, TIME INTERVALS
# time intervals = half years
lo_limitT = 0.1
# Convert the 'date' column to datetime type
vote_events = pd.read_csv(localpath + "data/vote_events.csv")
vote_events['date'] = pd.to_datetime(vote_events['date'])

# Get the first and last dates in the data
first_date = vote_events['date'].min()
last_date = vote_events['date'].max()

# Create a list to store the new DataFrames
new_dfs = []

# Iterate over each half year period
current_date = first_date
while current_date <= last_date:
  # Calculate the start and end dates of the current half year period
  period_start = pd.Timestamp(current_date.year, 1 if current_date.month <= 6 else 7, 1)
  period_end = pd.Timestamp(current_date.year, 7 if current_date.month <= 6 else 12, 31)

  # Filter the DataFrame for rows within the current half year period
  filtered_df = vote_events.loc[:, ['vote_event_id', 'date']].copy()

  filtered_df['in_time_period'] = filtered_df['date'].between(period_start, period_end)

  # Create a new variable 'in_time_period' indicating whether the row is in the given half year
  # filtered_df['in_time_period'] = True

  # Append the filtered DataFrame to the list of new DataFrames
  new_dfs.append(filtered_df)

  # Move to the next half year period
  current_date = period_end + pd.DateOffset(days=1)

# Access each new DataFrame in the list

projections = []
for i, filtered_df in enumerate(new_dfs):

  # filtered_df['in_time_period'].to_csv(localpath + "data/wpca_test.csv", index=False)

  XrawTc = Xraw.loc[:, selected_voters]
  filtered_df.index = filtered_df['vote_event_id']
  XrawTc[~filtered_df['in_time_period']] = np.nan



  # Xstand = Xraw.sub(Xraw.mean(axis=1), axis=0).div(Xraw.std(axis=1, ddof=0), axis=0)
  # XTc = (XrawTc - attr(Xstand,"scaled:center"))/attr(Xstand,"scaled:scale")
  XTc = Xraw.sub(Xraw.mean(axis=1), axis=0).div(Xraw.std(axis=1, ddof=0), axis=0)
  XTc[[~filtered_df['in_time_period']]] = np.nan

  # Indices of NAs; division x person
  TIc = XTc
  TIc = TIc.where(TIc.isna(), 1)
  TIc = TIc.fillna(0)
  TIc[[~filtered_df['in_time_period']]] = 0


  # weights for non missing data; divisions x persons
  # TIcw = TIc * w1 * w2
  # X = Xstand.mul(w1, axis=0).mul(w2, axis=0)
  TIcw = TIc.mul(w1, axis=0).mul(w2, axis=0)
  # sum of weights of divisions for each persons
  s = TIcw.sum(axis=0)
  pTW = s / max(s)
  # index of persons in calculation
  pTI = pTW > lo_limitT


  selected_votersT = pTI[pTI].index.tolist()

  XTw0cc = XTc.mul(w1, axis=0).mul(w2, axis=0).loc[:, selected_votersT].fillna(0)

  aZ = abs(Z)

  # dweights = t(t(aZ)%*%TIcc / apply(aZ,2,sum))      
  # dweights[is.na(dweights)] = 0

  TIcc = TIc.loc[:, selected_votersT]
  aZ_transpose = aZ.T

  column_sums = aZ.sum(axis=0)

  division_result = aZ_transpose / column_sums

  dweights = (aZ_transpose.dot(TIcc).T).div(column_sums).fillna(0)

  # XTw0ccproj = t(XTw0cc)%*%Z / dweights
  XTw0ccproj = XTw0cc.T.dot(Z).div(dweights)

  projections.append(XTw0ccproj)


# quick save to proper format
wpca1 = pd.read_csv(localpath + "data/wpca.v1.all.csv")

out = pd.DataFrame(columns=(wpca1.columns.tolist() + ['datum']))

for i, projection in enumerate(projections):
  p = projection.iloc[:, 0:3]
  p.columns = ['dim1', 'dim2', 'dim3']
  p.reset_index(inplace=True)
  p = p.merge(wpca1.loc[:, ['voter_id', 'příjmení', 'jméno', 'kandidátka', 'klub']], on='voter_id', how='left')
  p['datum'] = i

  out = pd.concat([out, p], axis=0)

# add no club
out['klub'] = out['klub'].fillna('Nezařazení')

# correct datum
def get_text(input_value):
  period_tf = {True: first_half_year['half'], False: 2 - ((first_half_year['half'] + 1) % 2)}
  year = first_half_year["year"] + (input_value + first_half_year["half"] - 1) // 2
  period = period_tf[(input_value % 2) == 0]
  return f"{period}. pol. {year}"

out['datum'] = out['datum'].apply(get_text)

# rotate
row = out.loc[out[rotate['column']] == rotate['value']]
# dims
for i in range(0, 2):
  if (rotate['dims'][i]) * row['dim' + str(i + 1)].values[0] < 0:
    out.loc[:, 'dim' + str(i + 1)] = out['dim' + str(i + 1)] * -1

out.to_csv(localpath + "data/wpca.halfyear.v1.csv", index=False)

File: wpca.r
# INPUT PARAMETERS
# _X_RAW_DB, _LO_LIMIT_1
# raw data in csv using db structure, i.e., a single row contains:
# code of representative, code of division, encoded vote (i.e. one of -1, 1, 0, NA)
# for example:
# “Joe Europe”,”Division-007”,”yes”


Xsource = read.csv("praha/data/votes.csv")

Xsource$vote_event_id = as.factor(Xsource$vote_event_id)
Xsource$voter_id = as.factor(Xsource$voter_id)
#Xsource = Xsource[c("voter_id","vote_event_id","option")]
Xsource$option_numeric = rep(0,length(Xsource$option))
Xsource$option_numeric[Xsource$option=='yes'] = 1
Xsource$option_numeric[Xsource$option=='no'] = -1
Xsource$option_numeric[Xsource$option=='abstain'] = -1
Xsource$option_numeric[Xsource$option=='not voting'] = NA
Xsource$option_numeric[Xsource$option=='absent'] = NA
Xsource$option_numeric = as.numeric(Xsource$option_numeric)

#Xrawdb = _X_RAW_DB
Xrawdb = Xsource
# lower limit to eliminate from calculations, e.g., .1; number
lo_limit = .1

# reorder data; divisions x persons
# we may need to install and/or load some additional libraries
# install.packages("reshape2")
library("reshape2")
# library("reshape2", lib.loc="/home/michal/R/x86_64-pc-linux-gnu-library/3.6")
# install.packages("sqldf")
# library("sqldf")

#prevent reordering, which is behaviour of acast:
#Xrawdb$V1 = factor(Xrawdb$V1, levels=unique(Xrawdb$V1))
Xrawdb$voter_id = factor(Xrawdb$voter_id, levels=unique(Xrawdb$voter_id))
Xraw = t(acast(Xrawdb,voter_id~vote_event_id,fun.aggregate=mean,value.var='option_numeric'))
Xpeople = dimnames(Xraw)[[2]]
Xvote_events = dimnames(Xraw)[[1]]
# Xraw=apply(Xraw,1,as.numeric)
mode(Xraw) = 'numeric'
# scale data; divisions x persons (mean=0 and sd=1 for each division)
Xstand=t(scale(t(Xraw),scale=TRUE))

# WEIGHTS
# weights 1 for divisions, based on number of persons in division
w1 = apply(abs(Xraw)==1,1,sum,na.rm=TRUE)/max(apply(abs(Xraw)==1,1,sum,na.rm=TRUE))
w1[is.na(w1)] = 0
# weights 2 for divisions, "100:100" vs. "195:5"
w2 = 1 - abs(apply(Xraw==1,1,sum,na.rm=TRUE) - apply(Xraw==-1,1,sum,na.rm=TRUE))/apply(!is.na(Xraw),1,sum)
w2[is.na(w2)] = 0

# analytical charts for weights:
#plot(w1)
#plot(w2)
#plot(w1*w2)

# weighted scaled matrix; divisions x persons
X = Xstand * w1 * w2

# MISSING DATA
# index of missing data; divisions x persons
I = X
I[!is.na(X)] = 1
I[is.na(X)] = 0

# weighted scaled with NA substituted by 0; division x persons
X0 = X
X0[is.na(X)]=0

# EXCLUSION OF REPRESENTATIVES WITH TOO FEW VOTES (WEIGHTED)
# weights for non missing data; division x persons
Iw = I*w1*w2
# sum of weights of divisions for each persons; vector of length “persons”
s = apply(Iw,2,sum)
pw = s/(t(w1)%*%w2)
# index of persons kept in calculation; vector of length “persons”
pI = pw > lo_limit
# weighted scaled with NA->0 and cutted persons with too few weighted votes; division x persons
X0c = X0[,pI]
# index of missing cutted (excluded) persons with too few weighted votes; divisions x persons
Ic = I[,pI]
# indexes of cutted (excluded) persons with too few votes; divisions x persons
Iwc = Iw[,pI]

# “X’X” MATRIX
# weighted X’X matrix with missing values substituted and excluded persons; persons x persons
C=t(X0c)%*%X0c * 1/(t(Iwc)%*%Iwc) * (sum(w1*w1*w2*w2))
# substitution of missing data in "covariance" matrix (the simple way)
C0 = C
C0[is.na(C)] = 0

# DECOMPOSITION
# eigendecomposition
Xe=eigen(C0)
# W (rotation values of persons)
W = Xe$vectors
# projected divisions into dimensions
Xy=X0c%*%W

# analytical charts of projection of divisions and lambdas
#plot(Xy[,1],Xy[,2])
#plot(sqrt(Xe$values[1:10]))

# lambda matrix
sigma = sqrt(Xe$values)
sigma[is.na(sigma)] = 0
lambda = diag(sigma)
# unit scaled lambda matrix
lambdau = sqrt(lambda^2/sum(lambda^2))

# projection of persons into dimensions
Xproj = W%*%lambda
# scaled projection of persons into dimensions
Xproju = W%*%lambdau*sqrt(dim(W)[1])

# analytical chart
# plot(Xproj[,1],Xproj[,2])
# plot(Xproju[,1],Xproju[,2])

# lambda^-1 matrix
lambda_1 = diag(sqrt(1/Xe$values))
lambda_1[is.na(lambda_1)] = 0

# Z (rotation values of divisions)
Z = X0c%*%W%*%lambda_1

# analytical charts
# second projection
Xproj2 = t(X0c) %*% Z
# without missing values, they are equal:
#plot(Xproj[,1],Xproj2[,1])
#plot(Xproj[,2],Xproj2[,2])

Xproj3u = W%*%lambdau * sqrt(lambda[1,1])
Xy3u=X0c%*%W / sqrt(lambda[1,1])

# "correlation matrix" (for charting)
invSC0 = solve(diag(sqrt(diag(C0))))
Corr0 = invSC0 %*% C0 %*% invSC0
rownames(Corr0) = rownames(C0)
colnames(Corr0) = colnames(C0)
# save into JSON file:
# setwd("home/michal/dev/wpca/psp/cs-2017-2019/")
# library("jsonlite", lib.loc="/home/michal/R/x86_64-pc-linux-gnu-library/3.4")
# sink(file="correlation_from_r.json")
# serializeJSON(as.data.frame(Corr0))
# sink()

# SAVE
write.csv(data.frame(data.frame(Xpeople, pI)[pI, ], Xproju[,1:3]), "Xproju.csv")
# write.csv(data.frame(Xpeople, pI)[pI, ], "Xpeople.csv")
